### PR TITLE
Check for RepType when getting element length

### PR DESCRIPTION
**Pull request by jadams-tresys**
DAFFODIL-2596

### CODE REVIEW

**Review by stevedlawrence — APPROVED** [review]
+1 seems fine, but it seems we we're going down the path of changing more and more variables to be something like:

```scala
val someVariable =
  if (optRepTypeElement.isDefined) {
    optRepTypeElement.get.someVariable
  } else {
    // actual variable logic
  }
```
There's now 

Feels like it's going to be a game a whack-a-mole to find all the right places where this is logic is needed. Makes me wonder if maybe we need to doing the `optRepTypeElement.isDefined` check a level or two above before these variables are even accessed? Or maybe we need something like this:

```
val repElement = optRepTypeElement.getOrElse(this)
```

And then when something accesses one of those variables that needs this isDefined check, we can use this new variable, for example:

```
repElement.fixedLengthValue
repElement.maybeFixedLengthInBits
```

I'm not sure if that's actually a good idea or not. Maybe these checks are actually better done or belong in this variables that care about repType? Just throwing out an idea that might make it easier to avoid these issues in the future?

File: daffodil-test/src/test/resources/org/apache/daffodil/extensions/type_calc/inputTypeCalc.tdml [hunk 1]
```diff
@@ -180,6 +182,22 @@
     </tdml:infoset>
   </tdml:parserTestCase>
 
+  <tdml:parserTestCase name="InputTypeCalc_keysetValue_03"
```
Comment (mbeckerle) [inline daffodil-test/src/test/resources/org/apache/daffodil/extensions/type_calc/inputTypeCalc.tdml:185]: change description string. Add comments to the file explaining the test case.

File: daffodil-test/src/test/resources/org/apache/daffodil/extensions/type_calc/inputTypeCalc.tdml [hunk 2]
```diff
@@ -50,6 +50,8 @@
       </xs:complexType>
     </xs:element>
 
+    <xs:element name="keysetValue_03" type="tns:keysetValue_01" dfdl:lengthKind="implicit"/>
```
Comment (mbeckerle) [inline daffodil-test/src/test/resources/org/apache/daffodil/extensions/type_calc/inputTypeCalc.tdml:53]: So is this the workaround? Wherever you use a type where the type carries a dfdlx:repType you need this dfdl:lengthKind 'implicit' ?

How come the other tests in here didn't need this?

Please add doc/comments explaining this and include the DAFFODIL-NNNN ticket number about the issue.

File: daffodil-test/src/test/scala/org/apache/daffodil/extensions/TestInputTypeValueCalc.scala [hunk 3]
```diff
@@ -43,10 +43,12 @@ class TestInputTypeValueCalc {
   @Test def test_InputTypeCalc_keysetValue_00(): Unit = { runner.runOneTest("InputTypeCalc_keysetValue_00") }
   @Test def test_InputTypeCalc_keysetValue_01(): Unit = { runner.runOneTest("InputTypeCalc_keysetValue_01") }
   @Test def test_InputTypeCalc_keysetValue_02(): Unit = { runner.runOneTest("InputTypeCalc_keysetValue_02") }
+  @Test def test_InputTypeCalc_keysetValue_03(): Unit = { runner.runOneTest("InputTypeCalc_keysetValue_03") }
```
Comment (mbeckerle) [inline daffodil-test/src/test/scala/org/apache/daffodil/extensions/TestInputTypeValueCalc.scala:46]: The name of this test probably should change also, as I don't think of this issue as a "inputTypeCalc" nor "keysetValue" thing. Suggestion: test_repTypeOnTypeRequiresLengthKindImplicit

**Review by mbeckerle — CHANGES_REQUESTED** [review]
Needs more doc, some test renaming.

**Comment by mbeckerle** [comment]
re: comment by @stevedlawrence on proliferating conditionals, the alternative is some object-oriented refactoring, where things that have repType get a different class than things without repType. This would actually require delegation I think to another object since we don't/can't analyze the repType property when the original DSOM object is created. I.e., we create the ElementBase instance as part of core parsing of the DFDL schema. Then, based on combining of element with its type, we eventually find out that it has a repType or not, and we'd have to construct a distinct object for the repType vs. Not case, and delegate the fixedLengthValue and elementLengthInBitsEv methods on that object. This is do-able, and may in fact be the right approach vs. proliferating conditionals of repType or not.

File: daffodil-test/src/test/resources/org/apache/daffodil/extensions/type_calc/inputTypeCalc.tdml (hunk 2 above)
Comment (jadams-tresys) [reply daffodil-test/src/test/resources/org/apache/daffodil/extensions/type_calc/inputTypeCalc.tdml:53]: Will do.  Not a workaround, merely setting it to implicit to trigger the bug that the code changes fixed.  The problem had been that it was thinking that the keysetValue_01 type was a string, which when lengthKind="implicit" causing it to complain about a missing maxLength property as you can't have an implicit length string without one.

This was one of the things we were missing when we were trying to debug the P8 schema yesterday.  The tests that were working had a lengthKind="delimited", which was fine given that it thought it was processing a string.  Adding dfdl:lengthKind="implicit" seemed like the easiest way to trigger the bug we were seeing in the P8 schema.

I'll work on adding some comments and a better name for the test now.

**Review by mbeckerle — APPROVED** [review]
+1.

**Comment by stevedlawrence** [comment]
What changed since this was last reviewed? Unfortunately GitHub doesn't have a way to show that.

Also there were some discussions about potentially switching to a OOP/delegation system to reduce all the if statements scattered throughout. Do we want to take that approach, or is that going to be too much effort for not much gain?

**Comment by jadams-tresys** [comment]
> What changed since this was last reviewed? Unfortunately GitHub doesn't have a way to show that.
> 
> Also there were some discussions about potentially switching to a OOP/delegation system to reduce all the if statements scattered throughout. Do we want to take that approach, or is that going to be too much effort for not much gain?

The change to OOP/delegation is what changed from the last commit.  Instead of adding more
`if (optRepTypeElement.isDefined)`
statements throughout the code I am know using the repElement suggestion you mentioned earlier in the review, which can either be the optRepTypeElement, or the element itself.

File: daffodil-core/src/main/scala/org/apache/daffodil/dsom/ElementBase.scala [hunk 4]
```diff
@@ -693,23 +693,19 @@ trait ElementBase
...
-        if (bitsMultiplier > 0) {
-          MaybeULong(fixedLengthValue * bitsMultiplier)
-        } else {
-          MaybeULong.Nope
-        }
+    if (isRepresented && repElement.isFixedLength) {
+      val bitsMultiplier = repElement.lengthUnits match {
+        case LengthUnits.Bits => 1
+        case LengthUnits.Bytes => 8
+        case LengthUnits.Characters => if (knownEncodingIsFixedWidth) knownEncodingWidthInBits else -1
```
Comment (stevedlawrence) [inline daffodil-core/src/main/scala/org/apache/daffodil/dsom/ElementBase.scala:700]: Should this be `repElement.knownEncoding...`

**Review by stevedlawrence — APPROVED** [review]
+1 This is a definite improvement.

It still feels a bit fragile in that there's a bunch of repElements scattered throughout and it's not obvious that we added all the places where it's need. But I guess if we're missing any they'll eventually show up as bugs, and if it continues we can consider some other refactoring.

**Pull request merged**